### DISCUSSION TITLE
Server Sent Events implementation

### DISCUSSION THREAD

**Discussion by Matasx** [Kaliumhexacyanoferrat/GenHTTP#309] [Q&A] [answered]
Hello,
first of all thank you for this awesome library, it works really well and the configuration and coding is efortless and feels very natural, I love it. 💯 

Though I was not able to figure out how to implement SSE. What I need is to get a response stream and keep the connection open while the client is connected so that I can keep posting data to that open stream. I think returning Stream as the response won't do it as peeking to your implementation it only reads all content while the stream Read function returns any data and then it stops, so I does not wait for the stream to be completed - or does not expect the stream producer to be async. 

Could you please advise me on how to implement this, if at all possible? Is there a way to get better control over the actual response and get handle of the output stream that I can keep alive and regularly post data into? 

(For reference here is SSE spec: https://html.spec.whatwg.org/multipage/server-sent-events.html ; note if you don't have any familiarity with it, it is a one-way stream from server to client, not to be confused with websockets)

Thank you in advance!

**Comment by Matasx** [comment, marked as answer]
Ok after a day of struggling I think I have figured out the missing piece. 

I implemented this response content:
``` C#
    public class PushStreamContent : IResponseContent
    {
        private readonly Func<Stream, uint, Task> _writer;

        public PushStreamContent(Func<Stream, uint, Task> writer)
        {
            _writer = writer;
        }

        public ulong? Length => null;

        public ValueTask<ulong?> CalculateChecksumAsync() => ValueTask.FromResult<ulong?>(null);

        public async ValueTask WriteAsync(Stream target, uint bufferSize)
        {
            await _writer(target, bufferSize);
        }
    }
```

Example usage:
``` C#
        [ResourceMethod(path: "test")]
        public IResponseBuilder Test(IRequest request)
        {
            return request
                .Respond()
                .Type(new FlexibleContentType("text/event-stream"))
                .Content(new PushStreamContent(async (stream, bufferSize) =>
                {
                    using var writer = new StreamWriter(stream, System.Text.Encoding.UTF8, (int)bufferSize);
                    var i = 0;
                    while (true)
                    {
                        await writer.WriteLineAsync("data: " + (++i));
                        await writer.WriteLineAsync();
                        await writer.FlushAsync();
                        await Task.Delay(1000);
                    }
                }));
        }
```

Still let me know if this is a proper way. Thanks. 😄

**Reply by Kaliumhexacyanoferrat** [reply]
Ah sorry, did forget about this issue - will have a look later 😉

**Comment by Kaliumhexacyanoferrat** [comment]
Sorry for the late reply - we had our daughter in May, so mail notifications like this can get lost in the shuffle. Thanks for the praise too, I'm always happy when someone appreciates the approach.

Your implementation is perfectly fine. Within a `while (true)` scenario I would add a `catch (IOException)` block preventing clients disconnecting from generating server errors. When integrating such a functionality into the framework, I would add another layer of `IHandler` and `IHandlerBuilder`, so you could write:

        [ResourceMethod(path: "test")]
        public IHandlerBuilder Test(IRequest request)
        {
            return Push.Create(async (stream) => {
                 // your lambda
            });
        }

If you would like to add your functionality (e.g. to the `IO` module), I would love to see a PR, otherwise I will create a backlog item for this.

**Reply by Matasx** [reply]
Got it. Thanks a lot. I think you can go ahead and create a backlog item. I might or might not get to it. 
PS: enjoy the time with your doughter, not coding. 😉

**Comment by Fiahblade** [comment]
I know it's a old thead but the stream is currently closed by an IO exception. This async task doesn't automatically close when you start adding more logic. (eg. only write when there is data)

**Comment by Kaliumhexacyanoferrat** [comment]
@Matasx, @Fiahblade: Support for Server Sent Events will be available in GenHTTP 9, see https://github.com/Kaliumhexacyanoferrat/GenHTTP.Website/blob/master/content/documentation/content/handlers/server-sent-events.md

**Reply by Matasx** [reply]
Wow, that's super cool and well put together. Thank you!

**Reply by Matasx** [reply]
One thing that I had to consider in my implementation of SSE was ping events, e.g. every 10s. That way both server and client knows whether they are still connected or not. As without sending any data, the disconnect won't happen on TCP/HTTP level and the socket might hang on server side when the client disconnects. At least that was my observation.

**Reply by Kaliumhexacyanoferrat** [reply]
Yes, TCP does not have a keep-alive mechanism so you will only notice that the connection has gone when you try to send data. That's why websockets have a ping built-in. Also some proxies will kill your connection if you do not send any data once in a while. Adding an automatic keep-alive feature would be interesting but probably add a lot of complexity as this needs to be done in an additional thread that needs to be synchronized with the generator thread.

Anyways, I created https://github.com/Kaliumhexacyanoferrat/GenHTTP/issues/533 for this. Thanks for your input :)
